reference-implementation/include/emitc/types.h

```diff
@@ -292,6 +292,34 @@ inline Dest binary(const SrcLeft &x, const SrcRight &y, BinaryOp &&op) {
   return z;
 }
 
+template <typename Dest, typename SrcA, typename SrcB, typename SrcC,
+          typename TernaryOp, IsScalar<SrcA> = true, IsScalar<SrcB> = true,
+          IsScalar<SrcC> = true>
+inline Dest ternary(const SrcA &a, const SrcB &b, const SrcB &c,
+                    TernaryOp &&op) {
+  return op(a, b, c);
+}
+
+template <typename Dest, typename SrcA, typename SrcB, typename SrcC,
+          typename TernaryOp, IsTensor<SrcA> = true, IsTensor<SrcB> = true,
+          IsTensor<SrcC> = true>
+inline Dest ternary(const SrcA &a, const SrcB &b, const SrcB &c,
+                    TernaryOp &&op) {
+  Dest d;
+  auto first1 = a.begin(), last1 = a.end();
+  auto first2 = b.begin(), first3 = c.begin();
+  auto result = d.begin();
+
+  while (first1 != last1) {
+    *result = op(*first1, *first2, *first3);
+    result++;
```

Comment:
Nit: prefer pre-increment unless post-increment behavior is needed.

(almost feels like this could have simpler with llvm::zip to bring all together, especially as you probably need to assert that we haven't hit the end of any of these).

Reply:
I was thinking the same thing, but I wasn't entirely sure how to wire llvm::zip into here.